posts/2019-01-17-Rust-1.32.0.md

```diff
@@ -0,0 +1,274 @@
+really remove it without causing a regression for people who do need
+jemalloc.
+
+Also, saying that `jemalloc` was always the default is a bit UNIX-centric,
+as it was only the default on *some* platforms. Notably, the MSVC target on
+Windows has shipped the system allocator for a long time.
+
+Finally, while jemalloc *usually* has great performance, that's not always
+the case. Additionally, it adds about 300kb to every Rust binary. We've also
+had a host of [other
+issues](https://github.com/rust-lang/rust/issues/36963#issuecomment-252029017)
+with jemalloc in the past. It has also felt a little strange that a systems
+language does not default to the system's allocator.
+
+For all of these reasons, once [Rust 1.28 shipped a way to choose a global
+allocator](https://blog.rust-lang.org/2018/08/02/Rust-1.28.html#whats-in-1.28.0-stable),
+we started making plans to switch the default to the system allocator, and
+allow you to use `jemalloc` via a crate. In Rust 1.32, we've finally finished
+this work, and by default, you will get the system allocator for your
+programs.
+
+If you'd like to continue to use jemalloc, use [the jemallocator crate]. In
+your `Cargo.toml`:
+
+```toml
+jemallocator = "0.1.8"
+```
+
+And in your crate root:
+
+```rust
+#[global_allocator]
+static ALLOC: jemallocator::Jemalloc = jemallocator::Jemalloc;
+```
+
+That's it! If you don't need jemalloc, it's not forced upon you, and if
+you do need it, it's a few lines of code away.
+
+[jemalloc]: http://jemalloc.net/
+[the jemallocator crate]: https://crates.io/crates/jemallocator
+
+#### Macro improvements
+
+A few improvements to macros have landed in Rust 1.32.0. First, [a new
+`literal` matcher](https://github.com/rust-lang/rust/pull/56072/) was added:
+
+```rust
+macro_rules! m {
+    ($lt:literal) => {};
+}
+
+fn main() {
+    m!("some string literal");
+}
+```
+
+`literal` mactches against literals of any type; string literals, numeric literals, `char` literals.
+
+In the 2018 edition, `macro_rules` macros can also use `?`, like this:
+
+```rust
+macro_rules! bar {
+    ($(a)?) => {}
+}
+```
+
+The `?` will match zero or one repetitions of the pattern, similar to the
+already-existing `*` for "zero or more" and `+` for "one or more."
+
+#### Final module improvements
+
+In Rust 1.30.0, [we announced several improvements to the module
+system](https://blog.rust-lang.org/2018/12/06/Rust-1.31-and-rust-2018.html#module-system-changes).
+We have one last tweak landing in 1.32.0 and the 2018 edition. Nicknamed
+["uniform paths"](https://github.com/rust-lang/rust/pull/56759/), the simplest way to explain
+it is to show you this code:
+
+```rust
+enum Color { Red, Green, Blue }
+
+use Color::*;
+```
+
+This code did *not* previously compile, as `use` statements had to start with
+`super`, `self`, or `crate`. With this change, this code will work, and do
+what you probably expect: import the variants of the `Color` enum defined
+above the `use` statement.
```

Comment:
Does the uniform-path stabilization _change_ the behavior of any code? The single-sentence description linked above is:

 > If import path resolves without errors, then it's resolved in exactly same way as a non-import path.

Is that summary _supposed_ to mean "more import paths now resolve without errors, and these paths are resolved in exactly the same way as non-import paths"?

Reply:
Okay, then, I'd recommend taking out the "simplest way to explain is..." phrasing, and say:

 > Nicknamed ["uniform paths"](link), it permits previously invalid import path statements to be resolved exactly the same way as non-import paths. For example:
>
> (enum example code)
>
> This code did *not* (etc)..., or `crate`. Now that the compiler supports uniform paths, the `use` statement will bring all variants of the `Color` enum into scope.